tests/json.test

```diff
@@ -0,0 +1,105 @@
+source [file dirname [info script]]/testing.tcl
+
+needs cmd json::decode json
+needs cmd json::encode json
+
+set json {
+{
+	"fossil":"9c65b5432e4aeecf3556e5550c338ce93fd861cc",
+	"timestamp":1435827337,
+	"command":"timeline/checkin",
+	"procTimeUs":3333,
+	"procTimeMs":3,
+	"homepage":null,
+	"payload":{
+		"limit":1,
+		"timeline":[{
+			"type":"checkin",
+			"uuid":"f8b17edee7ff4f16517601c40eb713602aed7a52",
+			"isLeaf":true,
+			"timestamp":1435318826,
+			"user":"juef",
+			"comment":"adwaita-icon-theme: update to 3.17.3",
+			"parents":["de628be645cc62429d630f9234c56d1fddfdc2a3"],
+			"tags":["trunk"]
+		}]
+	}
+}}
+
+test json-decode-001 {top level keys} {
+	lsort [dict keys [json::decode $json]]
+} {command fossil homepage payload procTimeMs procTimeUs timestamp}
+
+# Note that the decode will return the keys/values in order
+test json-decode-002 {object value} {
+	dict get [json::decode $json] payload
+} {limit 1 timeline {{type checkin uuid f8b17edee7ff4f16517601c40eb713602aed7a52 isLeaf true timestamp 1435318826 user juef comment {adwaita-icon-theme: update to 3.17.3} parents de628be645cc62429d630f9234c56d1fddfdc2a3 tags trunk}}}
+
+test json-decode-003 {object nested value} {
+	dict get [json::decode $json] payload timeline
+} {{type checkin uuid f8b17edee7ff4f16517601c40eb713602aed7a52 isLeaf true timestamp 1435318826 user juef comment {adwaita-icon-theme: update to 3.17.3} parents de628be645cc62429d630f9234c56d1fddfdc2a3 tags trunk}}
+
+test json-decode-004 {array entry from nested value} {
+	lindex [dict get [json::decode $json] payload timeline] 0
+} {type checkin uuid f8b17edee7ff4f16517601c40eb713602aed7a52 isLeaf true timestamp 1435318826 user juef comment {adwaita-icon-theme: update to 3.17.3} parents de628be645cc62429d630f9234c56d1fddfdc2a3 tags trunk}
+
+test json-decode-005 {object value from child array entry} {
+	dict get [lindex [dict get [json::decode $json] payload timeline] 0] comment
+} {adwaita-icon-theme: update to 3.17.3}
+
+test json-decode-006 {unicode escape} {
+	dict get [json::decode {{"key":"\u2022"}}] key
+} \u2022
+
+test json-decode-011 {null subsitution} {
+	dict get [json::decode -null NULL $json] homepage
+} {NULL}
+
+test json-decode-012 {default null value} {
+	dict get [json::decode $json] homepage
+} {null}
+
+unset -nocomplain json
+
+test json-encode-1.1 {String with backslashes}  {
+	json::encode {A "quoted string containing \backslashes\"}
+} {"A \"quoted string containing \\backslashes\\\""}
+
+test json-encode-1.2 {String with special chars} {
+	json::encode "Various \n special \b characters \t and /slash/ \r too"
+} {"Various \n special \b characters \t and \/slash\/ \r too"}
+
+test json-encode-1.3 {Array of numbers} {
+	json::encode {1 2 3.0 4} {list num}
+} {[ 1, 2, 3.0, 4 ]}
```

Comment:
I'd add more examples of floats.  
  -3.0
  3e0
  3.0e10
  3.0e-10
etc.

How are Inf and NaN handled? Does -0.0 preserve the negative sign?

Reply:
Good point. I'll add some more tests. Everything except Inf/Infinity "just works" because we just keep the value as-is. I'll add conversions for Inf <-> Infinity